### PR TITLE
Improved replacement of ephemerally-sourced symbols

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135635

**TL;DR:** This PR does the following:
* Uses `CleanDiv` instead of `FloorDiv` in more places when we know it holds to get strictly more information:
    * In `_try_isolate_lhs()`, if we know that `lhs` and `rhs` are integers, we can solve via `CleanDiv` instead of `/`
        * The PR teaches `_try_isolate_lhs()` how to handle `CleanDiv(numer, denom)` where the `denom` component contains the thing we're trying to isolate. This happens in practice for `slice()` and `view()` views
    * In the `infer_size()` util used by various views (including `view()` and `reshape()`), use `CleanDiv` for inferring the `-1` dimension from the `total size / product of other dims`
        * We also learn that all components of the product are >= 1, so provide that information as well

**Background:** During subclass fake-ification, subclass instances that are views go through the process of "view replay" to reconstruct the base -> view relationship in fake land. Part of view replay involves symbolicizing any closed-over ints in the original view function. For example, a slice view `x[3]` will have a `3` baked in as an arg to slice in the saved view func; we want to symbolicize this 3 for dynamic shapes support.

In practice, before the view is replayed, all closed-over ints are replaced with SymInts with "ephemeral sources" that are intended to disappear when we [assert](https://github.com/pytorch/pytorch/blob/49e0b88aab0fd124c318fe1c59fbbd8726298338/torch/_subclasses/meta_utils.py#L1073-L1078) the subclass size / stride / storage offset match the allocated, non-ephemeral outer symbolic values. For this to work, we rely on the [SymInt replacement logic](https://github.com/pytorch/pytorch/blob/49e0b88aab0fd124c318fe1c59fbbd8726298338/torch/fx/experimental/symbolic_shapes.py#L4740-L4744) (e.g. replace `s1` with `s0 * 5` if `Eq(s1, s0*5)` is found to be true), and [prioritize ephemerally-sourced SymInts](https://github.com/pytorch/pytorch/blob/49e0b88aab0fd124c318fe1c59fbbd8726298338/torch/fx/experimental/symbolic_shapes.py#L4909-L4921) to be replaced first.

**Problem:** In certain cases, the SymInt replacement logic does not apply, and an ephemerally-sourced SymInt lives for longer than originally designed, resulting in guard time errors like those reported [here](https://github.com/pytorch/pytorch/pull/133337#issuecomment-2302417084) and in #128649.
```
AssertionError: s7 (could be from ['<ephemeral: symint_visitor_fn>', '<ephemeral: symint_visitor_fn>']) not in {s3: ["L['nt']._values.size()[0]", "L['nt']._values.size()[0]"], s0: ["L['nt'].size()[1]"], s6: ["L['nt'].size()[2]", "L['nt']._values.size()[1]"], s1: ["L['nt']._values.size()[0]"], s7: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

**Solution:** For each of the error cases, we can identify why a particular SymInt replacement is not made and enhance the logic to support such replacements. This PR addresses this for `slice()` and `view()` views that have been encountered in practice. For example:
```python
@torch.compile(backend="eager", dynamic=True)
def f(t):
    return t._base + 1

x_a = torch.randn(4, 4, requires_grad=True)
x = TwoTensor(x_a, x_a.clone())
out = f(x[3])
```
The input to the compiled func is a subclass view produced by `slice()`. Fake-ification of this subclass view does the following:
1. Allocates non-ephemeral size / stride / storage offset symbols for the outer subclass metadata. Since the slice is contiguous with shape `(4)`, we get `(s0)` for the shape, `(1)` for the stride, and `s1` for the storage offset.
2. Perform view replay on a fake-ified base to reconstruct the base -> view relationship. The captured `3` arg to `slice()` is replaced with an ephemerally-sourced SymInt `s2` before the view func is replayed. The output of view replay has shape `(s0)`, stride `(1)`, and `s0 * s2` for the storage offset.
3. [Asserts](https://github.com/pytorch/pytorch/blob/49e0b88aab0fd124c318fe1c59fbbd8726298338/torch/_subclasses/meta_utils.py#L1073-L1078) that the output of view replay has the shape metadata allocated in step (1). Since we have `Eq(s1, s0 * s2)` from the storage offset assert and `s2` is ephemeral, we should replace `s2` with `CleanDiv(s1, s0)`. Before this PR, this replacement was not supported since `s1/s0` is not guaranteed to be an integer. With `CleanDiv`, we have the integer guarantee and the replacement can happen.
https://github.com/pytorch/pytorch/blob/49e0b88aab0fd124c318fe1c59fbbd8726298338/torch/fx/experimental/symbolic_shapes.py#L4969-L4970

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by ezyang** [comment]
This fails tests. Do you want review anyway?

**Comment by ezyang** [comment]
One basic comment here: since the output of this division is guaranteed to be integral, why aren't we using CleanDiv here?

**Comment by jbschlosser** [comment]
> This fails tests. Do you want review anyway?

I'll have some time today to look at the test failures. As far as review goes, any high-level insight on the right way to address this class of errors is appreciated :)

> One basic comment here: since the output of this division is guaranteed to be integral, why aren't we using CleanDiv here?

This seems reasonable, let me look into this!

**Comment by jbschlosser** [comment]
@ezyang I updated the PR to use `CleanDiv`, which is in fact fairly clean. This addresses the ephemeral source error for the `slice()` and `view()` cases described in https://github.com/pytorch/pytorch/pull/133337#issuecomment-2302417084 and https://github.com/pytorch/pytorch/issues/128649 by enhancing the symbol replacement logic to work in more places.

One notable case still fails for some reason: `view(-1)` with `torch.compile(..., dynamic=True)`. cc @bdhirsh if you have any insight.
```python
import torch
from torch.testing._internal.two_tensor import TwoTensor

def f(x):
    return x * 2

f_compiled = torch.compile(f, backend="aot_eager", dynamic=True)

a, b = (torch.randn(4, 15) for _ in range(2))
t = TwoTensor(a, b)

t_view = t.view(-1)
out_ref = f(t_view)
out_test = f_compiled(t_view)
```
```
AssertionError: s0 (could be from ["L['x']._base.size()[0]", "L['x']._base.a.size()[0]", "L['x']._base.b.size()[0]"]) not in {}.
```

File: torch/_prims_common/__init__.py [hunk 1]
```diff
@@ -921,7 +921,31 @@ def infer_size(shape: ShapeType, numel: int) -> Tuple[int, ...]:
...
+
+        from torch.fx.experimental.sym_node import SymNode
+        from torch.utils._sympy.functions import CleanDiv
+
         shape[dim] = numel // newsize
+        if isinstance(newsize, torch.SymInt):
+            # we now know that every component of the newsize multiplication must be >= 1
+            if newsize.node.shape_env is not None:
+                newsize.node.shape_env._constrain_range_for_size(
+                    newsize.node.expr, min=1, max=None
```
Comment (ezyang) [inline torch/_prims_common/__init__.py:933]: you don't have to do this. Just do `torch._check(newsize >= 1)`

File: torch/_prims_common/__init__.py [hunk 2]
```diff
@@ -921,7 +921,31 @@ def infer_size(shape: ShapeType, numel: int) -> Tuple[int, ...]:
...
+                )
+                for s in newsize.node.expr.args:
+                    newsize.node.shape_env._constrain_range_for_size(s, min=1, max=None)
+
+            # Specify that this is a CleanDiv and not just a FloorDiv
```
Comment (ezyang) [inline torch/_prims_common/__init__.py:938]: This is also not the right way to do this. I think what you should do is assert exactly divisibility (torch._check(numel % newsize == 0)) and then the division should automatically convert into a CleanDiv, although I haven't tried this recently so I don't know if it will work out of the box.

File: torch/utils/_sympy/solve.py [hunk 3]
```diff
@@ -117,8 +117,12 @@ def _try_isolate_lhs(
         # That is because we don't know whether we have mirror the operation or not.
         if not (isinstance(e, INEQUALITY_TYPES) and other.is_negative is None):
             # Divide both sides by 'other'.
-            lhs = lhs / other
-            rhs = rhs / other
+            if lhs.is_integer and rhs.is_integer and other.is_integer:
+                lhs = CleanDiv(lhs, other)
+                rhs = CleanDiv(rhs, other)
```
Comment (ezyang) [inline torch/utils/_sympy/solve.py:122]: Can you explain why this is OK to do for relational tests? This is only obviously sound when e is a sympy.Eq

File: torch/utils/_sympy/solve.py [hunk 4]
```diff
@@ -128,6 +132,26 @@ def _try_isolate_lhs(
             assert op is not None
             e = op(lhs, rhs)
 
+    ################################################################################
+    # left-hand side is CleanDiv
+    ################################################################################
+    if (
```
Comment (ezyang) [inline torch/utils/_sympy/solve.py:138]: Can you write a comment similar to 

```
    # Given the expression: a // b op c
```

in the other cases

Comment (ezyang) [reply torch/utils/_sympy/solve.py:138]:
```
CleanDiv(numer, denom) == rhs

=>

denom == CleanDiv(numer, rhs)
```

File: torch/utils/_sympy/solve.py [hunk 5]
```diff
@@ -128,6 +132,26 @@ def _try_isolate_lhs(
...
+        and e.lhs.has(thing)
+    ):
+        numer, denom = e.lhs.args
+        thing_in_numer = numer.has(thing)
+        thing_in_denom = denom.has(thing)
+
+        # only support thing in denominator for now
+        if thing_in_denom and thing_in_numer != thing_in_denom:
+            return _try_isolate_lhs(
+                sympy.Eq(denom, CleanDiv(numer, e.rhs)), thing, floordiv_inequality
```
Comment (ezyang) [inline torch/utils/_sympy/solve.py:152]: Potential infinite loop when rhs contains thing.

```
CleanDiv(8, G(thing)) = F(thing)
```

you then simplify to 

```
G(thing) = CleanDiv(8, F(thing))
```

but this will trigger this logic again.

Do you still have enough juice if you exclude thing from rhs?

**Comment by jbschlosser** [comment]
I will revisit this but after discussion with @bdhirsh indicating there's no real-world use case atm I've deprioritized it a bit. Feel free to yell at me if you want the priority bumped back up :)

**Comment by bdhirsh** [comment]
In particular:

(1) nobody is technically directly blocked by this right now

(2) it may show up at any time though with DTensor, since DTensor has separate outer_size/stride symbols. Someone will run into it if they compile a model that takes a view of a DTensor as a graph input, and that view closes over sizes that become the actual size of the view (e.g. indexing, like `view = inp[3]`)

(3) it also shows up in the AOTAutograd tests when we update `TwoTensor` to have separate outer sizes (although we can xfail the test for now)
